mobsf/MobSF/security.py

```diff
@@ -249,3 +252,67 @@ def sanitize_for_logging(filename: str, max_length: int = 255) -> str:
 
     # Truncate filename to the maximum allowed length
     return filename[:max_length]
+
+
+def valid_host(host):
+    """Check if host is valid, run SSRF checks."""
+    default_timeout = socket.getdefaulttimeout()
+    try:
+        if len(host) > 2083:  # Standard URL length limit
+            return False
+
+        prefixs = ('http://', 'https://')
+        if not host.startswith(prefixs):
+            host = f'http://{host}'
+        parsed = urlparse(host)
+        domain = parsed.netloc
+        hostname = parsed.hostname
+        path = parsed.path
+        query = parsed.query
+        params = parsed.params
+
+        # Check for hostname
+        if not hostname:
+            return False
+
+        # Check for URL credentials
+        if '@' in domain:
+            return False
+
+        # Detect parser escapes, only host is allowed
+        if len(path) > 0 or len(query) > 0 or len(params) > 0:
+            return False
+
+        # Resolve dns to get ipv4 or ipv6 address
+        socket.setdefaulttimeout(5)  # 5 second timeout
+        ip_addresses = socket.getaddrinfo(hostname, None)
```

Comment:
Each time this is called a DNS rebinding is possible

Reply:
Internal port scanning is still possible with the next urls, the `ip_addresses = socket.getaddrinfo(hostname, None)` does issue a DNS query thus permitting the host to switch from `1.1.1.1:80` to `127.0.0.1:80`

- "prefix1-make-1.1.1.1-rebindfor30safter1times-127.0.0.1-rr.1u.ms:80"
- "prefix1-make-1.1.1.1-rebindfor30safter1times-127.0.0.1-rr.1u.ms:81"
- ....
- "prefix1-make-1.1.1.1-rebindfor30safter1times-127.0.0.1-rr.1u.ms:443"

measuring the response time would yield if the port is open or not.

Reply:
Since firebase and assetlink looks at https://, or http://, I can also do a port check.

Reply:
how about using the initially retrieved value of `socket.getaddreinfo` to avoid the DNS rebind ?

Reply:
We cannot rely on the IP alone to resolve a resource. It depends on the virtual hosting or domain based routing configurations of the endpoint.

a single IP address can host different resources at the same path across multiple domain names.

Reply:
indeed, but after the resolving of `socket.getaddrinfo()` the next call to `.get()` would point to the second DNS rebinded host, thus still vulnerable to a certain extent with a limited impact.

Regards

Reply:
Yes, I acknowledge that this PR does not address the DNS rebind issue. A potential fix could involve monkey patching requests/urllib3 or implementing a custom HTTP request library with fixed domain name and IP. However, I am hesitant to add this unless there is a medium/high impact real-world exploit case. The current checks in place offer a balanced approach to mitigating the SSRF issue and its primary exploit use cases.

Reply:
I see now, anyway, keep me posted about when you would release the advisory. Thank you